Added ACI-CNI 5.2.3.6 support

siva-muni opened the pull request:
https://github.com/rancher/rancher/issues/41124

LavanyaVaddavalli commented:
During the rke binary generation for our testing, one of the steps is to generate a pseudo-version using 'go get....rke-commit-id' , and place the pseudo-version in go.mod of kdm repo for 'go generate' and 'go mod tidy'. The 'go mod tidy' is failing with the below error. We ignored this error, and used the data.json generated in 'go generate' step to generate the rke binary and used it for testing. We are not sure if we should be concerned about this issue.

```go: finding module for package k8s.io/apimachinery/pkg/util/clock
go: downloading k8s.io/apimachinery v0.27.0
github.com/rancher/kontainer-driver-metadata/rke imports
        github.com/rancher/rke/types imports
        github.com/rancher/norman/condition imports
        github.com/rancher/norman/controller imports
        k8s.io/client-go/tools/cache imports
        k8s.io/apimachinery/pkg/util/clock: module k8s.io/apimachinery@latest found (v0.27.0), but does not contain package k8s.io/apimachinery/pkg/util/clock```

LavanyaVaddavalli commented:
Can you please review the PR @kinarashah @snasovich

snasovich reviewed (changes requested):
@siva-muni @LavanyaVaddavalli , thank you for submitting this PR.

We would like to align the release of updated ACI-CNI with April 2023 patches we're tracking in https://github.com/rancher/rancher/milestone/311 milestone.

More specifically, new versions will be added per https://github.com/rancher/rancher/issues/41054, so realistically we want new ACI-CNI versions to be updated on to-be-added `v1.24.13-rancher1-1` and `v1.25.9-rancher1-1`.
There is an existing "in review" PR that adds them (https://github.com/rancher/kontainer-driver-metadata/pull/1115) - however note the target branch for it will be changed to https://github.com/rancher/kontainer-driver-metadata/tree/dev-v2.7-2023-04-patches.

So I recommend developing changes with the assumption we want to add your changes to versions to be added per https://github.com/rancher/rancher/issues/41054.

Also note there won't be new 1.23 version, so it's fine to introduce `v1.23.16-rancher2-2` but the comment should be stating `// Enabled out of band post v2.7.2`.

And finally, is this only getting addressed for 2.7.x or there is intention to get it into 2.6.x as well?

snasovich commented:
> During the rke binary generation for our testing, one of the steps is to generate a pseudo-version using 'go get....rke-commit-id' , and place the pseudo-version in go.mod of kdm repo for 'go generate' and 'go mod tidy'. The 'go mod tidy' is failing with the below error. We ignored this error, and used the data.json generated in 'go generate' step to generate the rke binary and used it for testing. We are not sure if we should be concerned about this issue.
> 
> ```
> go: downloading k8s.io/apimachinery v0.26.3
> github.com/rancher/kontainer-driver-metadata/rke imports
>         github.com/rancher/rke/types imports
>         github.com/rancher/norman/condition imports
>         github.com/rancher/norman/controller imports
>         k8s.io/client-go/tools/cache imports
>         k8s.io/apimachinery/pkg/util/clock: module k8s.io/apimachinery@latest found (v0.26.3), but does not contain package k8s.io/apimachinery/pkg/util/clock```
> ```

@kinarashah , could you please review and comment on this?

siva-muni commented:
> @siva-muni @LavanyaVaddavalli , thank you for submitting this PR.
> 
> We would like to align the release of updated ACI-CNI with April 2023 patches we're tracking in https://github.com/rancher/rancher/milestone/311 milestone.
> 
> More specifically, new versions will be added per [rancher/rancher#41054](https://github.com/rancher/rancher/issues/41054), so realistically we want new ACI-CNI versions to be updated on to-be-added `v1.24.13-rancher1-1` and `v1.25.9-rancher1-1`. There is an existing "in review" PR that adds them (#1115) - however note the target branch for it will be changed to https://github.com/rancher/kontainer-driver-metadata/tree/dev-v2.7-2023-04-patches.
> 
> So I recommend developing changes with the assumption we want to add your changes to versions to be added per [rancher/rancher#41054](https://github.com/rancher/rancher/issues/41054).
> 
> Also note there won't be new 1.23 version, so it's fine to introduce `v1.23.16-rancher2-2` but the comment should be stating `// Enabled out of band post v2.7.2`.
> 
> And finally, is this only getting addressed for 2.7.x or there is intention to get it into 2.6.x as well?

Thanks @snasovich, addressed the comments in b634fe03332c2d9ccbdf8b40e0bf033031ee5e1e and raised PR against dev-v2.7-2023-04-patches branch

Now we added 3 kubernetes versions,
v1.23.16-rancher2-2, with the starting comment updated to "// Enabled out of band post v2.7.2."
v1.24.13-rancher1-1, took this version from PR(https://github.com/rancher/kontainer-driver-metadata/pull/1115) and added aci 5.2.3.6 images
v1.25.9-rancher1-1, took this version from PR (https://github.com/rancher/kontainer-driver-metadata/pull/1115) and added aci 5.2.3.6 images


- We will do backport to 2.6.x as well, for KDM changes will do for the branch [v2.6-KDM-April-2023-patches](https://github.com/rancher/rancher/milestone/312).  rke repo changes will do for branch release/v1.3

siva-muni commented:
Testing in-progress after changing the target branches. Once testing done, will let you know.

siva-muni commented:
unable to test v1.24.13-rancher1-1, v1.25.9-rancher1-1 kubernetes versions since some of the system images like rancher/hyperkube:v1.24.13-rancher1 and  rancher/hyperkube:v1.25.9-rancher1 seems to have not been pushed into [docker.io](http://docker.io/) yet.

snasovich reviewed (changes requested):
With https://github.com/rancher/kontainer-driver-metadata/pull/1115 now merged, please re-base this PR.

Otherwise the changes LGTM.

siva-muni commented:
thanks [snasovich](https://github.com/snasovich) for the review. I Rebased this PR

kinarashah commented:
> During the rke binary generation for our testing, one of the steps is to generate a pseudo-version using 'go get....rke-commit-id' , and place the pseudo-version in go.mod of kdm repo for 'go generate' and 'go mod tidy'. The 'go mod tidy' is failing with the below error. We ignored this error, and used the data.json generated in 'go generate' step to generate the rke binary and used it for testing. We are not sure if we should be concerned about this issue.
> 
> ```
> go: downloading k8s.io/apimachinery v0.27.0
> github.com/rancher/kontainer-driver-metadata/rke imports
>         github.com/rancher/rke/types imports
>         github.com/rancher/norman/condition imports
>         github.com/rancher/norman/controller imports
>         k8s.io/client-go/tools/cache imports
>         k8s.io/apimachinery/pkg/util/clock: module k8s.io/apimachinery@latest found (v0.27.0), but does not contain package k8s.io/apimachinery/pkg/util/clock```
> ```

@LavanyaVaddavalli The reason this error is showing up is because we're bumping RKE to a significantly newer version which also requires pinning apimachinery. Without pinning to a particular version, it pulls the latest v0.27.0 and complains about a missing package. (https://github.com/rancher/rke/blob/v1.4.4/go.mod#L37) This can be resolved by pinning `k8s.io/apimachinery` and `k8s.io/client-go` in `replace` section. 
RKE is vendored into KDM only for the types definition, and the new fields in RKE PR aren't referenced by KDM directly in the system image struct, it should be okay to ignore the error as you did for this case because it'll have no net effect on the generated data.json. But in future, if we add any new fields to structs being used in KDM, we'll need to update RKE and address errors accordingly.

kinarashah commented on rke/k8s_rke_system_images.go:
```diff
@@ -8374,6 +8374,50 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
 			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.6",
 			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.21.1",
 		},
+		// Enabled out of band post v2.7.1
```
Nit: this needs to be `post v2.7.2`

siva-muni replied:
Addressed in https://github.com/rancher/kontainer-driver-metadata/pull/1110/commits/5b25798542eab0b4cdb306179109f8d58d8ac3dc

siva-muni commented:
@kinarashah, @snasovich  
Addressed all the comments, please re-review.

snasovich commented:
@siva-muni , please note that we're bumping versions a bit per https://github.com/rancher/kontainer-driver-metadata/pull/1123 - this PR will need to be rebased again once that one is merged (should be today) and template version ranges changed slightly.

kinarashah commented:
@siva-muni as @snasovich mentioned above, the linked PR is merged so ready to rebase this PR and update version range to say `rancher2`.

siva-muni committed:
Added ACI-CNI support for 5.2.3.6

siva-muni committed:
go generate

siva-muni commented:
> @siva-muni as @snasovich mentioned above, the linked PR is merged so ready to rebase this PR and update version range to say `rancher2`.

@kinarashah, Please note that we rebased the PR and updated the version range with `rancher2`

snasovich reviewed (approved):
LGTM. @kinarashah , please review and merge if it looks good for you.
